### PR TITLE
Add LNaddress to lnurlp

### PR DESCRIPTION
**Pull request by bitkarrot**
Work in progress, add Lightning address to lnurlp

### CODE REVIEW

**Comment by bitkarrot** [comment]
Ok this PR is ready for review, someone please review it. 

Thank you!

**Comment by bitkarrot** [comment]
@callebtc or someone please to review! thank you

File: templates/lnurlp/index.html [hunk 1]
```diff
@@ -56,6 +57,18 @@ <h5 class="text-subtitle1 q-my-none">Pay links</h5>
...
+                  dense
+                  size="xs"
+                  icon="checkmark"
+                  :color="($q.dark.isActive) ? 'grey-7' : 'grey-5'"
+                  type="a"
+                  :href="'https://sendsats.to/' + props.row.username +'@' + domain"
+                  target="_blank"
+                >
+                </q-btn>
+                <q-tooltip>Check LN Address works via external call</q-tooltip>
```
Comment (callebtc) [inline templates/lnurlp/index.html:71]: I think this is unnecessary and could potentially be a privacy issues. Can you get rid of this button?

File: crud.py [hunk 2]
```diff
@@ -1,12 +1,49 @@
...
+# from loguru import logger
+
+async def check_lnaddress_update(username: str, id: str) -> bool:
+    # check no duplicates for lnaddress when updating an username
+    row = await db.fetchall(
+        "SELECT username FROM lnurlp.pay_links WHERE username = ? AND id = ?",
+        (username, id),
+    )
+    if len(row) > 1:
+        assert False, "Lightning Address Already exists. Try a different One?"
```
Comment (callebtc) [inline crud.py:18]:
```suggestion
        assert False, "Username already exists. Try a different one."
```

File: crud.py [hunk 3]
```diff
@@ -1,12 +1,49 @@
...
+        return True
+
+
+async def check_lnaddress_exists(username: str) -> bool:
+    # check if lnaddress username exists in the database when creating a new entry
+    row = await db.fetchall(
+        "SELECT username FROM lnurlp.pay_links WHERE username = ?", (username,)
+    )
+    if row:
+        assert False, "Lighting Address Already exists. Try a different One?"
```
Comment (callebtc) [inline crud.py:30]:
```suggestion
        assert False, "Username already exists. Try a different one."
```

File: lnurl.py [hunk 4]
```diff
@@ -1,16 +1,77 @@
...
+async def api_lnurl_callback(request: Request, link_id, amount: int = Query(...)):
+    address = await get_pay_link(link_id)
+    if not address:
+        return LnurlErrorResponse(reason=f'{"Address not found"}').dict()
+
+    domain = urlparse(str(request.url)).netloc
+    assert domain
+
+    unhashed_description = await address.lnurlpay_metadata(domain=domain)
+    unhashed_description = unhashed_description.encode()
```
Comment (callebtc) [inline lnurl.py:52]: Type changes mid-flight. Better: 
```suggestion
    unhashed_description = await address.lnurlpay_metadata(domain=domain).encode()
```

File: lnurl.py [hunk 5]
```diff
@@ -1,16 +1,77 @@
...
+    payment_hash, payment_request = await create_invoice(
+        wallet_id=address.wallet,
+        amount=int(amount / 1000),
+        memo=address.description,
+        unhashed_description=unhashed_description,
+        extra={
+            "tag": "lnurlp",
+            "link": address.id,
+            "extra": {"tag": f"Payment to {address.username}@{domain}"},
```
Comment (callebtc) [inline lnurl.py:61]: This extra should not be here. Could be a non-ln address payment as well? In either case, this should not be here.

File: lnurl.py [hunk 6]
```diff
@@ -1,16 +1,77 @@
...
+    return resp
+
+
+# for lnaddress callback
+@lnurlp_ext.get(
+    "/api/v1/lnurl/cb/{link_id}",
+    status_code=HTTPStatus.OK,
+    name="lnurlp.api_lnurl_callback",
+)
+async def api_lnurl_callback(request: Request, link_id, amount: int = Query(...)):
```
Comment (callebtc) [inline lnurl.py:43]: I just noticed that a function with the same name exists in the same file at line 113. Does this even work?

**Comment by callebtc** [comment]
The extensions errors when you have old LNURLp links because of pydantic validation errors. The username must be declared as optional in `models.py` (L45). Please also add a Query param (L26).

The callback function is duplicate in `lnurl.py`. IMO, it should be integrated into the original function and not be a separate function. Wouldn't that be clearner?

**Pull request closed without merging**